cpp/src/quantiles/tdigest/tdigest.cu

```diff
@@ -0,0 +1,356 @@
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/tdigest/tdigest.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/structs/structs_column_view.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+// https://developer.nvidia.com/blog/lerp-faster-cuda/
+template <typename T>
+__device__ inline T lerp(T v0, T v1, T t)
+{
+  return fma(t, v1, fma(-t, v0, v0));
+}
+
+template <typename T>
+struct cast_output {
+  data_type t;
+
+  __device__ device_storage_type_t<T> operator()(double v)
+  {
+    if constexpr (!cudf::is_fixed_point<T>()) { return static_cast<T>(v); }
+    if constexpr (cudf::is_fixed_point<T>()) {
+      return T{v, numeric::scale_type{t.scale()}}.value();
+    }
+  }
+};
+
+// kernel for computing percentiles on input tdigest (mean, weight) centroid data.
+template <typename T>
+__global__ void compute_percentiles_kernel(offset_type const* tdigest_offsets,
+                                           size_type num_tdigests,
+                                           device_span<double const> percentiles,
+                                           double const* mean_,
+                                           double const* weight_,
+                                           double const* min_,
+                                           double const* max_,
+                                           double const* cumulative_weight_,
+                                           device_storage_type_t<T>* output,
+                                           data_type output_type)
+{
+  int const tid = threadIdx.x + blockIdx.x * blockDim.x;
+
+  auto const tdigest_index = tid / percentiles.size();
+  if (tdigest_index >= num_tdigests) { return; }
+  double const percentage = percentiles[tid % percentiles.size()];
+
+  // size of the digest we're querying
+  auto const tdigest_size = tdigest_offsets[tdigest_index + 1] - tdigest_offsets[tdigest_index];
+  if (tdigest_size == 0) {
+    // no work to do. values will be set to null
+    return;
+  }
+  double const* cumulative_weight = cumulative_weight_ + tdigest_offsets[tdigest_index];
+
+  // means and weights for the tdigest
+  double const* mean   = mean_ + tdigest_offsets[tdigest_index];
+  double const* weight = weight_ + tdigest_offsets[tdigest_index];
+
+  // min and max for the digest
+  double const* min_val = min_ + tdigest_index;
+  double const* max_val = max_ + tdigest_index;
+
+  double const total_weight = cumulative_weight[tdigest_size - 1];
+  double const cluster_q    = percentage * total_weight;
+
+  cast_output<T> cast_op{output_type};
+
+  if (cluster_q <= 1) {
+    output[tid] = cast_op(*min_val);
+    return;
+  } else if (cluster_q >= total_weight - 1) {
+    output[tid] = cast_op(*max_val);
+    return;
+  }
```

Comment:
This feels like a comment or a refactor to better express intent of what's going on here would be helpful. It could be as simple as giving names to the condtions `cluster_q <= 1` and `cluster_q >= total_weight - 1`.